arch/risc-v/src/common/riscv_pmp.c

```diff
@@ -0,0 +1,206 @@
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/arch.h>
+
+#include "riscv_internal.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define PMP_CFG_BITS_CNT        (8)
+#define PMP_CFG_FLAG_MASK       (0xFF)
+
+#define PMP_CFG_CNT_IN_REG      (__riscv_xlen / PMP_CFG_BITS_CNT)
+
+#define PMP_MASK_SET_ONE_REGION(region, attr, reg) \
+  do { \
+      uintptr_t offset = region % PMP_CFG_CNT_IN_REG; \
+      reg &= ~(PMP_CFG_FLAG_MASK << (offset * PMP_CFG_BITS_CNT)); \
+      reg |= attr << (offset * PMP_CFG_BITS_CNT); \
+    } while(0);
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: riscv_config_pmp_region
+ *
+ * Description:
+ *   This function will set the specific PMP region with the desired cfg.
+ *
+ * Input Parameters:
+ *   region - The region index number.
+ *   attr - The region configurations.
+ *   base - The base address of the region.
+ *   size - The memory length of the region.
+ *   For the NAPOT mode, the base address must aligned to the size boundary,
+ *   and the size must be power-of-two according to the the PMP spec.
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+void riscv_config_pmp_region(uintptr_t region, uintptr_t attr,
+                             uintptr_t base, uintptr_t size)
+{
+  uintptr_t addr = 0;
+  uintptr_t cfg = 0;
+
+  /* TODO: check the base address alignment and size */
+
+  addr = base >> 2;
+  if (PMPCFG_A_NAPOT == (attr & PMPCFG_A_MASK))
+    {
+      addr |= (size - 1) >> 3;
+    }
+
+  switch (region)
+    {
+      case 0:
+        __asm volatile("csrw pmpaddr0, %0" : :"r"(addr));
```

Comment:
change all csrw and csrr to WRITE_CSR and READ_CSR?

Reply:
OK，let me update them all.Thx.